From d59cd7edd78901db8fd88c86258f6830e35e91df Mon Sep 17 00:00:00 2001
From: Yu Watanabe <watanabe.yu+github@gmail.com>
Date: Mon, 25 Feb 2019 14:56:21 +0800
Subject: [PATCH 28/30] musl: glob-util: filter out . and .. even if
 GLOB_ALTDIRFUNC is not supported

musl neither support GLOB_ALTDIRFUNC nor GLOB_BRACE.
Let's make safe_glob() work even when GLOB_ALTDIRFUNC is not supported.
Currently, GLOB_BRACE is simply ignored when it is not supported.
---
 src/basic/glob-util.c     | 75 +++++++++++++++++++++++++++++++++++++--
 src/basic/glob-util.h     |  9 ++++-
 src/include/musl/glob.h   | 21 +++++++++++
 src/test/test-glob-util.c | 10 ++++++
 4 files changed, 112 insertions(+), 3 deletions(-)
 create mode 100644 src/include/musl/glob.h

diff --git a/src/basic/glob-util.c b/src/basic/glob-util.c
index 5843ef088f..84b369be87 100644
--- a/src/basic/glob-util.c
+++ b/src/basic/glob-util.c
@@ -9,9 +9,42 @@
 #include "string-util.h"
 #include "strv.h"
 
+static bool safe_glob_verify(const char *p, const char *prefix) {
+        if (isempty(p))
+                return false; /* should not happen, but for safey. */
+
+        if (prefix) {
+                /* Skip the prefix, as we allow dots in prefix.
+                 * Note, glob() does not normalize paths, hence do not use path_startswith(). */
+                p = startswith(p, prefix);
+                if (!p)
+                        return false; /* should not happen, but for safety. */
+        }
+
+        for (;;) {
+                p += strspn(p, "/");
+                if (*p == '\0')
+                        return true;
+                if (*p == '.') {
+                        p++;
+                        if (IN_SET(*p, '/', '\0'))
+                                return false; /* refuse dot */
+                        if (*p == '.') {
+                                p++;
+                                if (IN_SET(*p, '/', '\0'))
+                                        return false; /* refuse dot dot */
+                        }
+                }
+
+                p += strcspn(p, "/");
+                if (*p == '\0')
+                        return true;
+        }
+}
+
 DEFINE_TRIVIAL_DESTRUCTOR(closedir_wrapper, void, closedir);
 
-int safe_glob_full(const char *path, int flags, opendir_t opendir_func, char ***ret) {
+int safe_glob_internal(const char *path, int flags, bool use_gnu_extension, opendir_t opendir_func, char ***ret) {
         _cleanup_(globfree) glob_t g = {
                 .gl_closedir = closedir_wrapper,
                 .gl_readdir = (struct dirent* (*)(void *)) readdir_no_dot,
@@ -23,8 +56,16 @@ int safe_glob_full(const char *path, int flags, opendir_t opendir_func, char ***
 
         assert(path);
 
+        // TODO: expand braces if GLOB_BRACE is specified but not supported.
+
+#if GLOB_ALTDIRFUNC == 0
+        use_gnu_extension = false;
+#else
+        SET_FLAG(flags, GLOB_ALTDIRFUNC, use_gnu_extension);
+#endif
+
         errno = 0;
-        r = glob(path, flags | GLOB_ALTDIRFUNC, NULL, &g);
+        r = glob(path, flags, NULL, &g);
         if (r == GLOB_NOMATCH)
                 return -ENOENT;
         if (r == GLOB_NOSPACE)
@@ -32,6 +73,36 @@ int safe_glob_full(const char *path, int flags, opendir_t opendir_func, char ***
         if (r != 0)
                 return errno_or_else(EIO);
 
+        if (!use_gnu_extension) {
+                _cleanup_free_ char *prefix = NULL;
+                r = glob_non_glob_prefix(path, &prefix);
+                if (r < 0 && r != -ENOENT)
+                        return r;
+
+                _cleanup_strv_free_ char **filtered = NULL;
+                size_t n_filtered = 0;
+                STRV_FOREACH(p, g.gl_pathv) {
+                        if (!safe_glob_verify(*p, prefix))
+                                continue;
+
+                        if (!ret)
+                                return 0; /* Found at least one entry, let's return earlier. */
+
+                        /* When musl is used, each entry is not a head of allocated memory. Hence, it is
+                         * necessary to copy the string. */
+                        r = strv_extend_with_size(&filtered, &n_filtered, *p);
+                        if (r < 0)
+                                return r;
+                }
+
+                if (n_filtered == 0)
+                        return -ENOENT;
+
+                assert(ret);
+                *ret = TAKE_PTR(filtered);
+                return 0;
+        }
+
         if (strv_isempty(g.gl_pathv))
                 return -ENOENT;
 
diff --git a/src/basic/glob-util.h b/src/basic/glob-util.h
index ea3e869319..36bfd6716c 100644
--- a/src/basic/glob-util.h
+++ b/src/basic/glob-util.h
@@ -7,7 +7,14 @@
 
 typedef DIR* (*opendir_t)(const char *);
 
-int safe_glob_full(const char *path, int flags, opendir_t opendir_func, char ***ret);
+int safe_glob_internal(const char *path, int flags, bool use_gnu_extension, opendir_t opendir_func, char ***ret);
+static inline int safe_glob_test(const char *path, int flags, char ***ret) {
+        /* This is for testing the fallback logic for the case GLOB_ALTDIRFUNC is not supported. */
+        return safe_glob_internal(path, flags, false, NULL, ret);
+}
+static inline int safe_glob_full(const char *path, int flags, opendir_t opendir_func, char ***ret) {
+        return safe_glob_internal(path, flags, true, opendir_func, ret);
+}
 static inline int safe_glob(const char *path, int flags, char ***ret) {
         return safe_glob_full(path, flags, NULL, ret);
 }
diff --git a/src/include/musl/glob.h b/src/include/musl/glob.h
new file mode 100644
index 0000000000..58e6c50678
--- /dev/null
+++ b/src/include/musl/glob.h
@@ -0,0 +1,21 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include_next <glob.h>
+
+/* Here, we set 0 to GLOB_ALTDIRFUNC and GLOB_BRACE, rather than the values used by glibc,
+ * to indicate that glob() does not support these flags. */
+
+#ifndef GLOB_ALTDIRFUNC
+#define GLOB_ALTDIRFUNC 0
+#define gl_flags    __dummy1
+#define gl_closedir __dummy2[0]
+#define gl_readdir  __dummy2[1]
+#define gl_opendir  __dummy2[2]
+#define gl_lstat    __dummy2[3]
+#define gl_stat     __dummy2[4]
+#endif
+
+#ifndef GLOB_BRACE
+#define GLOB_BRACE 0
+#endif
diff --git a/src/test/test-glob-util.c b/src/test/test-glob-util.c
index a9880f15c8..730961d172 100644
--- a/src/test/test-glob-util.c
+++ b/src/test/test-glob-util.c
@@ -61,20 +61,30 @@ TEST(safe_glob) {
 
         fn = strjoina(template, "/*");
         ASSERT_ERROR(safe_glob(fn, /* flags = */ 0, &v), ENOENT);
+        ASSERT_ERROR(safe_glob_test(fn, /* flags = */ 0, &v), ENOENT);
 
         fn2 = strjoina(template, "/.*");
         ASSERT_ERROR(safe_glob(fn2, GLOB_NOSORT|GLOB_BRACE, &v), ENOENT);
+        ASSERT_ERROR(safe_glob_test(fn2, GLOB_NOSORT|GLOB_BRACE, &v), ENOENT);
 
         fname = strjoina(template, "/.foobar");
         ASSERT_OK(touch(fname));
 
         ASSERT_ERROR(safe_glob(fn, /* flags = */ 0, &v), ENOENT);
+        ASSERT_ERROR(safe_glob_test(fn, /* flags = */ 0, &v), ENOENT);
 
         ASSERT_OK(safe_glob(fn2, GLOB_NOSORT|GLOB_BRACE, &v));
         ASSERT_EQ(strv_length(v), 1u);
         ASSERT_STREQ(v[0], fname);
         ASSERT_NULL(v[1]);
 
+        v = strv_free(v);
+
+        ASSERT_OK(safe_glob_test(fn2, GLOB_NOSORT|GLOB_BRACE, &v));
+        ASSERT_EQ(strv_length(v), 1u);
+        ASSERT_STREQ(v[0], fname);
+        ASSERT_NULL(v[1]);
+
         (void) rm_rf(template, REMOVE_ROOT|REMOVE_PHYSICAL);
 }
 
-- 
2.51.0

